tasks/libs/pipeline_tools.py

```diff
@@ -62,6 +62,19 @@ def cancel_pipelines_with_confirmation(gitlab, pipelines):
             print(f"Pipeline {color_message(pipeline['id'], 'bold')} will keep running.\n")
 
 
+def gracefully_cancel_pipeline(gitlab, pipeline):
+    """
+    Gracefully cancel pipeline
+    - Cancel all the jobs that did not start to run yet
+    - Do not cancel jobs containing 'cleanup' in their name
+    """
+    jobs = gitlab.all_jobs(pipeline["id"])
+
+    for job in jobs:
+        if job["status"] not in ["running", "canceled"] and "cleanup" not in job["name"]:
```

Comment:
Couldn't we cancel jobs in the `binary_build` stage as well? They are fairly long and shouldn't require any cleanup

An exception might be the macOS build job that runs on github, but this might come at a later point

Reply:
That's interesting! I am not confident enough to affirm that cancelling jobs in `binary_build` cannot have any bad impacts but I would be really happy to add this to this PR. Because indeed those jobs take a very long time to run

Reply:
The proposed solution works for me, thanks, and nice future proofing if we realize a new stage can be cancelled :)